File: perl/bf.pl
#!/usr/bin/env perl
use strict;
use warnings;

local $| = 1;
my @a = (0) x 30000;
my $p = 0;
my %translation = (
    '>' => '++$p',
    '<' => '--$p',
    '+' => '++$a[$p] == 256 and $a[$p] = 0',
    '-' => '--$a[$p] == -1 and $a[$p] = 255',
    '.' => 'print chr($a[$p])',
    ',' => '$a[$p] = ord(getc) unless eof(STDIN)',
    '[' => 'while ($a[$p]) {',
    ']' => '}',
);

my $program = do {
    local $/ = undef;
    <>;
};
die $! unless defined($program);

$program =~ s/[^><+-.,[\]]//g;
$program =~ s/./$translation{$&};/g;
eval $program;
die $@ if $@;

File: ruby/bf.rb
#!/usr/bin/env ruby
a = Array.new(30_000, 0)
p = 0
translation = {
  '>' => 'p += 1;',
  '<' => 'p -= 1;',
  '+' => 'a[p] = a[p]+1 & 255;',
  '-' => 'a[p] = a[p]-1 & 255;',
  '.' => 'putc a[p].chr;',
  ',' => 'a[p] = STDIN.getbyte unless STDIN.eof;',
  '[' => 'while a[p].nonzero? do ',
  ']' => 'end;'
}

eval ARGF.read.gsub(/./, translation)
